<a href="https://colab.research.google.com/github/kayurh/data_analytics/blob/main/1_1_Lecture_Kayur_Amour_Harry.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1 lecture – Structured Data and DataFrame Handling (Part 1)
- The goal of this lesson is to get familiar with the concept of a Pandas DataFrame and how to use it.
- You will learn how to load, modify, and analyze data using a DataFrame.
- By the end of the lesson, students will be able to perform simple data manipulation tasks in Python.

### Introduction

What is a DataFrame?
-  Tabular data structure (similar to an Excel sheet or an SQL table).
-  It consists of rows and columns, with indexes and labels.

Why is a DataFrame important in Data Science?
-  An efficient way to store and manipulate data.
-  Easy to read, transform, and visualize.

Required Python packages
- pandas, numpy, matplotlib
- Installation (if necessary): pip install pandas numpy matplotlib


### Extra / Further reading

English video about DataFrames: [click here](https://www.youtube.com/watch?v=eQ0xPt6hIfs)


### Creating a DataFrame from Lists and Dictionaries

Imagine we are building a small database of a company’s employees.

How can we convert this into a DataFrame?

In [ ]:
# Import the pandas package (commonly used with the alias 'pd')
import pandas as pd

# Create a dictionary where each key maps to a list
# - 'Name' column: names of three people
# - 'Age' column: ages
# - 'City' column: places of residence
data = {'name': ['Anna', 'Béla', 'Csaba'],
        'age': [23, 45, 32],
        'city': ['Budapest', 'Szeged', 'Debrecen']}

In [ ]:
display(data)

{'name': ['Anna', 'Béla', 'Csaba'],
 'age': [23, 45, 32],
 'city': ['Budapest', 'Szeged', 'Debrecen']}

In [ ]:
# A pandas DataFrame is used to store data in a tabular (table-like) format.
# Here, the dictionary keys become the column names, and the list elements become the row values.
df = pd.DataFrame(data)

In [ ]:
# Print the DataFrame.
# (In a Jupyter notebook, 'display(df)' shows the result in a nicer table format.)
display(df)

,name,age,city
0,Anna,23,Budapest
1,Béla,45,Szeged
2,Csaba,32,Debrecen


Creating a DataFrame from Lists

In [ ]:
# Store the data in three separate lists
names = ['Anna', 'Béla', 'Csaba']          # list of names
age = [23, 45, 32]                       # list of ages
city = ['Budapest', 'Szeged', 'Debrecen']  # list of cities

In [ ]:
print(names)
print(age)
print(city)

['Anna', 'Béla', 'Csaba']
[23, 45, 32]
['Budapest', 'Szeged', 'Debrecen']


In [ ]:
# The zip() function "packs" the elements of the lists together:
# [('Anna', 23, 'Budapest'),
#  ('Béla', 45, 'Szeged'),
#  ('Csaba', 32, 'Debrecen')]
# That is, each row becomes a tuple (name, age, city).
# list() converts the zip object into an actual list.
list(zip(names, age, city))

[('Anna', 23, 'Budapest'), ('Béla', 45, 'Szeged'), ('Csaba', 32, 'Debrecen')]

In [ ]:
# Then we load the data into a pandas DataFrame,
# and specify the column names using the 'columns' parameter.
df = pd.DataFrame(list(zip(names, age, city)),
                  columns=['name', 'age', 'city'])

In [ ]:
# In a Jupyter notebook, display() shows the data in a nice table format.
display(df)

,name,age,city
0,Anna,23,Budapest
1,Béla,45,Szeged
2,Csaba,32,Debrecen


Creating a DataFrame from a List of Lists

In [ ]:
# Create a list where each element is another list.
# Each inner list represents one data row: [name, age, city].
adatok = [
    ['Anna', 23, 'Budapest'],
    ['Béla', 45, 'Szeged'],
    ['Csaba', 32, 'Debrecen']
]

# When creating the DataFrame:
# - 'adatok' is interpreted as rows
# - column names are provided via the 'columns' parameter.
df = pd.DataFrame(adatok, columns=['name', 'age', 'city'])

# Print the DataFrame.
# (In Jupyter, display(df) is nicer.)
print(df)


    name  age      city
0   Anna   23  Budapest
1   Béla   45    Szeged
2  Csaba   32  Debrecen


Creating a DataFrame from a NumPy array

If we receive the data as a NumPy array (e.g., from machine learning models or mathematical operations),
we can easily convert it using the pandas.DataFrame() constructor.

In [ ]:
import numpy as np # Import NumPy (for arrays and numerical operations)

# Create a NumPy array.
# Each row contains one person's data: [name, age, city].
numpy_array = np.array([
    ['Anna', 23, 'Budapest'],
    ['Béla', 45, 'Szeged'],
    ['Csaba', 32, 'Debrecen']
])

# Pass the NumPy array to the pandas DataFrame constructor.
# The 'columns' parameter defines the column names.
df = pd.DataFrame(numpy_array, columns=['name', 'age', 'city'])


# Print the DataFrame.
# (In a notebook, it’s better to use display(df).)
print(df)


    name age      city
0   Anna  23  Budapest
1   Béla  45    Szeged
2  Csaba  32  Debrecen


Advantages:

If the source data comes as a NumPy array (e.g., ML datasets), this method is convenient.
It may be faster for large datasets.

### Creating / Loading a DataFrame from a CSV file

In real life, data is often stored in CSV files.
The pandas read_csv() function is one of the best tools for loading CSV data.

In [ ]:
# Read a CSV (Comma-Separated Values) file directly from the internet.
# pandas.read_csv() automatically downloads and loads the table as a DataFrame.
# The data can also come from any local file path.
df = pd.read_csv('https://gist.github.com/netj/8836201/raw/6f9306ad21398ea43cba4f7d537619d0e07d5ae3/iris.csv')

# head() shows the first 5 rows by default.
# This is useful for quickly checking what the table contains.
display(df.tail())  # display the first 5 rows

,sepal.length,sepal.width,petal.length,petal.width,variety
145,6.7,3.0,5.2,2.3,Virginica
146,6.3,2.5,5.0,1.9,Virginica
147,6.5,3.0,5.2,2.0,Virginica
148,6.2,3.4,5.4,2.3,Virginica
149,5.9,3.0,5.1,1.8,Virginica


Advantages:
- Easy when you receive data from an external source (e.g., Excel or a database).
- Pandas automatically handles missing values.
- The default separator is a comma (","). If the separator is different (e.g. ";" or tab),
  you must specify it.
  
More details: https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

### Loading a CSV with different parameters

In practice, CSV files can come in many different formats:
- they may use a different separator (e.g., ";" or tab instead of ",")
- they might not have a header, or the header may not be on the first row
- they may contain special characters, accents, or missing values marked in different ways

pandas.read_csv() provides many parameters to handle these cases.
Some commonly used settings:

- sep → field separator character
- header / names → header row index or custom column names
- index_col → which column should be used as the index
- usecols → load only specific columns
- dtype → enforce data types
- na_values → specify which values should be treated as missing
- skiprows, nrows → skip rows / read only a limited number of rows
- encoding → file encoding (important for special characters)

In the example below, we demonstrate how these parameters can be used.


In [ ]:
# CSV loading with useful parameters (example)
df = pd.read_csv(
    'https://gist.github.com/netj/8836201/raw/6f9306ad21398ea43cba4f7d537619d0e07d5ae3/iris.csv',  # file path or URL
    sep=",",               # column separator (e.g., ";" or "\t" are also possible)
    header=0,              # which row to use as the header (0 = first row)
    names=None,            # if there is no header, we can provide column names, e.g. ['A','B','C']
    index_col=None,        # which column should be used as the index (e.g., 'ID')
    usecols=None,          # load only specific columns, e.g. ['sepal.length','variety']
    dtype=None,            # enforce data types, e.g. {'sepal.length': float, 'variety': 'category'}
    na_values=["?", "NA"], # values to treat as missing data
    skiprows=0,            # number of rows to skip at the start of the file
    nrows=None,            # number of rows to read (e.g., for testing: nrows=100)
    encoding="utf-8"       # file encoding (sometimes needed e.g. "latin1" or "cp1250" for special characters)
)

# Quick check
print(df.head())



   sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa


In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal.length  150 non-null    float64
 1   sepal.width   150 non-null    float64
 2   petal.length  150 non-null    float64
 3   petal.width   150 non-null    float64
 4   variety       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB
None


### Summary


From a dictionary (dict)

In [ ]:
import pandas as pd

data_dict = {
    "Method": ["Dictionary (dict)", "List (list of lists)", "Lists with zip()",
               "NumPy array", "CSV file", "Excel file", "JSON file", "SQL database"],
    "Use case": ["Small datasets", "Tabular datasets", "Simple lists",
                 "Mathematical operations / numeric pipelines", "Real-world datasets", "Tabular datasets",
                 "Loading API data", "Large datasets"],
    "Advantages": ["Fast and easy to read", "Easy to convert", "Easy to combine with other data types",
                   "Fast processing for large datasets", "Easy integration with external data sources",
                   "Can handle multiple sheets", "Easy data exchange", "Efficient queries"]
}
# DataFrame
df_dict = pd.DataFrame(data_dict)
display(df_dict)


,Method,Use case,Advantages
0,Dictionary (dict),Small datasets,Fast and easy to read
1,List (list of lists),Tabular datasets,Easy to convert
2,Lists with zip(),Simple lists,Easy to combine with other data types
3,NumPy array,Mathematical operations / numeric pipelines,Fast processing for large datasets
4,CSV file,Real-world datasets,Easy integration with external data sources
5,Excel file,Tabular datasets,Can handle multiple sheets
6,JSON file,Loading API data,Easy data exchange
7,SQL database,Large datasets,Efficient queries


List of lists

In [ ]:
# Data in a list-of-lists format
data_list = [
    ["Dictionary (dict)", "For small datasets", "Fast and easy to read"],
    ["List (list of lists)", "Tabular data", "Easy to convert"],
    ["Lists using zip()", "For simple lists", "Easy to combine with other data types"],
    ["NumPy array", "Mathematical operations", "Fast processing for large datasets"],
    ["CSV file", "Real-world datasets", "Easy integration with other data sources"],
    ["Excel file", "Tabular data", "Can handle multiple sheets"],
    ["JSON file", "Loading API data", "Easy data exchange"],
    ["SQL database", "Large datasets", "Efficient queries"]
]

# Create a DataFrame
df_list = pd.DataFrame(data_list, columns=["Method", "Use case", "Advantages"])
display(df_list)


,Method,Use case,Advantages
0,Dictionary (dict),For small datasets,Fast and easy to read
1,List (list of lists),Tabular data,Easy to convert
2,Lists using zip(),For simple lists,Easy to combine with other data types
3,NumPy array,Mathematical operations,Fast processing for large datasets
4,CSV file,Real-world datasets,Easy integration with other data sources
5,Excel file,Tabular data,Can handle multiple sheets
6,JSON file,Loading API data,Easy data exchange
7,SQL database,Large datasets,Efficient queries


Using zip()

In [ ]:
# Store the data in separate lists
methods = ["Dictionary (dict)", "List (list of lists)", "Lists using zip()",
           "NumPy array", "CSV file", "Excel file", "JSON file", "SQL database"]

use_case = ["For small datasets", "Tabular data", "For simple lists",
            "Mathematical operations", "Real-world datasets", "Tabular data",
            "Loading API data", "Large datasets"]

advantages = ["Fast and easy to read", "Easy to convert", "Easy to combine with other data types",
              "Fast processing for large datasets", "Easy integration with other data sources",
              "Can handle multiple sheets", "Easy data exchange", "Efficient queries"]

# Create a DataFrame
df_zip = pd.DataFrame(zip(methods, use_case, advantages), columns=["Method", "Use case", "Advantages"])
display(df_zip)


,Method,Use case,Advantages
0,Dictionary (dict),For small datasets,Fast and easy to read
1,List (list of lists),Tabular data,Easy to convert
2,Lists using zip(),For simple lists,Easy to combine with other data types
3,NumPy array,Mathematical operations,Fast processing for large datasets
4,CSV file,Real-world datasets,Easy integration with other data sources
5,Excel file,Tabular data,Can handle multiple sheets
6,JSON file,Loading API data,Easy data exchange
7,SQL database,Large datasets,Efficient queries


From a NumPy array

In [ ]:
import numpy as np

# Create a NumPy array from the list-of-lists
data_numpy = np.array(data_list)

# Create a DataFrame from the NumPy array
df_numpy = pd.DataFrame(data_numpy, columns=["Method", "Use case", "Advantages"])
display(df_numpy)


,Method,Use case,Advantages
0,Dictionary (dict),For small datasets,Fast and easy to read
1,List (list of lists),Tabular data,Easy to convert
2,Lists using zip(),For simple lists,Easy to combine with other data types
3,NumPy array,Mathematical operations,Fast processing for large datasets
4,CSV file,Real-world datasets,Easy integration with other data sources
5,Excel file,Tabular data,Can handle multiple sheets
6,JSON file,Loading API data,Easy data exchange
7,SQL database,Large datasets,Efficient queries


### The IRIS dataset

The Iris dataset is a classic dataset in machine learning and data analysis.
It contains 150 flower samples from three different iris species (setosa, versicolor, virginica).

Each sample has four features:
- sepal.length (sepal length, cm)
- sepal.width (sepal width, cm)
- petal.length (petal length, cm)
- petal.width (petal width, cm)
- variety (iris species name)

More details: https://archive.ics.uci.edu/dataset/53/iris

Loading data

In [ ]:
import pandas as pd

# Load the Iris dataset
df = pd.read_csv('https://gist.github.com/netj/8836201/raw/6f9306ad21398ea43cba4f7d537619d0e07d5ae3/iris.csv')

# View the first 5 rows
display(df.head())



,sepal.length,sepal.width,petal.length,petal.width,variety
0,5.1,3.5,1.4,0.2,Setosa
1,4.9,3.0,1.4,0.2,Setosa
2,4.7,3.2,1.3,0.2,Setosa
3,4.6,3.1,1.5,0.2,Setosa
4,5.0,3.6,1.4,0.2,Setosa


Let’s check how many rows and columns the original DataFrame has

In [ ]:
df.shape

(150, 5)

In [ ]:
# no. rows
df.shape[0]

150

### Column selection

### Selecting a single column

If we want to access only one column, we can reference it by name:

In [ ]:
display(df["sepal.length"])


0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: sepal.length, Length: 150, dtype: float64

This returns a Pandas Series object.

### Selecting multiple columns

If we want to access multiple columns at once, we pass a list of column names:

In [ ]:
df_subset = df[["sepal.length", "sepal.width"]]
display(df_subset.head())


,sepal.length,sepal.width
0,5.1,3.5
1,4.9,3.0
2,4.7,3.2
3,4.6,3.1
4,5.0,3.6


### Selecting columns by position (iloc)

Sometimes we don’t know the column name, only its position. That’s what .iloc is for:

In [ ]:
display(df.iloc[:, 0])  # colselection
display(df.iloc[:, [0, 1]])  # Selecting a col by index (iloc)


0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: sepal.length, Length: 150, dtype: float64

,sepal.length,sepal.width
0,5.1,3.5
1,4.9,3.0
2,4.7,3.2
3,4.6,3.1
4,5.0,3.6
...,...,...
145,6.7,3.0
146,6.3,2.5
147,6.5,3.0
148,6.2,3.4


### Row selection

### Selecting a row by index (iloc)

With .iloc[] we can access rows by integer position:

In [ ]:
display(df.iloc[0])  # first row


sepal.length       5.1
sepal.width        3.5
petal.length       1.4
petal.width        0.2
variety         Setosa
Name: 0, dtype: object

Selecting multiple rows:

In [ ]:
display(df.iloc[0:5])  # the first 5 rows


,sepal.length,sepal.width,petal.length,petal.width,variety
0,5.1,3.5,1.4,0.2,Setosa
1,4.9,3.0,1.4,0.2,Setosa
2,4.7,3.2,1.3,0.2,Setosa
3,4.6,3.1,1.5,0.2,Setosa
4,5.0,3.6,1.4,0.2,Setosa


### Selecting a row by index label (loc)

If the data contains unique index labels (e.g. IDs or names), we can use loc[]:

In [ ]:
print(df.loc[0])  # Selecting rows by index


sepal.length       5.1
sepal.width        3.5
petal.length       1.4
petal.width        0.2
variety         Setosa
Name: 0, dtype: object


Selecting multiple rows by label:

In [ ]:
print(df.loc[[0, 1, 2]])  # show rows based on index

   sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa


In [ ]:
df['variety'].unique()

array(['Setosa', 'Versicolor', 'Virginica'], dtype=object)

 ### Filtering rows based on conditions

For example, if we want to query only the samples of the Setosa species:

In [ ]:
df_setosa = df[df["variety"] == "Setosa"]
print(df_setosa.head())


   sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa


Let’s check how many rows the original DataFrame has

In [ ]:
df.shape[0]

150

How many rows remain after filtering?

In [ ]:
df_setosa.shape[0]

50

In percentage?

In [ ]:
# rounded to 2 decimals
round((df_setosa.shape[0]/df.shape[0])*100,2)

33.33

If we want to apply multiple conditions
(e.g., only Setosa samples with sepal length > 5 cm):

In [ ]:
df_filtered = df[(df["variety"] == "Setosa") & (df["sepal.length"] > 5)]
print(df_filtered.head())


    sepal.length  sepal.width  petal.length  petal.width variety
0            5.1          3.5           1.4          0.2  Setosa
5            5.4          3.9           1.7          0.4  Setosa
10           5.4          3.7           1.5          0.2  Setosa
14           5.8          4.0           1.2          0.2  Setosa
15           5.7          4.4           1.5          0.4  Setosa


### Combined selection of rows and columns

If we want to select both specific rows and columns, we can combine .iloc[] or .loc[]:

In [ ]:
print(df.iloc[0:5, 0:2])


   sepal.length  sepal.width
0           5.1          3.5
1           4.9          3.0
2           4.7          3.2
3           4.6          3.1
4           5.0          3.6


If we want to select by column names:

In [ ]:
print(df.loc[0:5, ["sepal.length", "variety"]])


   sepal.length variety
0           5.1  Setosa
1           4.9  Setosa
2           4.7  Setosa
3           4.6  Setosa
4           5.0  Setosa
5           5.4  Setosa


### Checking unique values

What unique values appear in the "variety" column?

In [ ]:
print(df["variety"].unique())

['Setosa' 'Versicolor' 'Virginica']


What are we checking here?
- Do the data contain all expected categories?
- If we saw a species called "XYZ", that would be suspicious.

If we want to see the frequency of each category:

In [ ]:
print(df["variety"].value_counts())

variety
Setosa        50
Versicolor    50
Virginica     50
Name: count, dtype: int64


What are we checking here?
- We check whether the distribution is balanced.

### Renaming columns
Sometimes column names are not appropriate or are too long.

In [ ]:
df = df.rename(columns={"sepal.length": "sepal_length", "sepal.width": "sepal_width"})
display(df.head())


,sepal_length,sepal_width,petal.length,petal.width,variety
0,5.1,3.5,1.4,0.2,Setosa
1,4.9,3.0,1.4,0.2,Setosa
2,4.7,3.2,1.3,0.2,Setosa
3,4.6,3.1,1.5,0.2,Setosa
4,5.0,3.6,1.4,0.2,Setosa


### Creating a new column

In [ ]:
df['new'] = 0
df.head()

,sepal_length,sepal_width,petal.length,petal.width,variety,new
0,5.1,3.5,1.4,0.2,Setosa,0
1,4.9,3.0,1.4,0.2,Setosa,0
2,4.7,3.2,1.3,0.2,Setosa,0
3,4.6,3.1,1.5,0.2,Setosa,0
4,5.0,3.6,1.4,0.2,Setosa,0


Conditional new column creation

In [ ]:
df['size'] = 'medium'
df.loc[df['sepal_length'] > 5, 'size'] = 'large'
df.loc[df['sepal_length'] < 4.7, 'size'] = 'small'

df.head()


,sepal_length,sepal_width,petal.length,petal.width,variety,new,size
0,5.1,3.5,1.4,0.2,Setosa,0,large
1,4.9,3.0,1.4,0.2,Setosa,0,medium
2,4.7,3.2,1.3,0.2,Setosa,0,medium
3,4.6,3.1,1.5,0.2,Setosa,0,small
4,5.0,3.6,1.4,0.2,Setosa,0,medium


### Dropping columns

If we don’t need a column, we can delete it.

In [ ]:
df = df.drop(columns=["new","size"])
display(df.head())


,sepal_length,sepal_width,petal.length,petal.width,variety
0,5.1,3.5,1.4,0.2,Setosa
1,4.9,3.0,1.4,0.2,Setosa
2,4.7,3.2,1.3,0.2,Setosa
3,4.6,3.1,1.5,0.2,Setosa
4,5.0,3.6,1.4,0.2,Setosa


### Selecting all numeric columns

In [ ]:
# Select numeric columns
numeric_columns = df.select_dtypes(include=["number"]).columns
numeric_columns


Index(['sepal_length', 'sepal_width', 'petal.length', 'petal.width'], dtype='object')

### Selecting only non-numeric columns

If we want the columns that are not numeric:

In [ ]:
non_numeric_columns = df.select_dtypes(exclude=["number"]).columns
non_numeric_columns


Index(['variety'], dtype='object')

### Sorting in descending order
If we want to sort the dataset by sepal length in descending order:

In [ ]:
df_sorted = df.sort_values(by="sepal_length", ascending=False)
display(df_sorted.head())


,sepal_length,sepal_width,petal.length,petal.width,variety
131,7.9,3.8,6.4,2.0,Virginica
122,7.7,2.8,6.7,2.0,Virginica
118,7.7,2.6,6.9,2.3,Virginica
117,7.7,3.8,6.7,2.2,Virginica
135,7.7,3.0,6.1,2.3,Virginica


### Randomly shuffling the data

If the data is ordered, it may influence analysis and modeling.
That’s why we shuffle it




In [ ]:
df_shuffled = df.sample(frac=1, random_state=42).reset_index(drop=True)
display(df_shuffled.head())

,sepal_length,sepal_width,petal.length,petal.width,variety
0,6.1,2.8,4.7,1.2,Versicolor
1,5.7,3.8,1.7,0.3,Setosa
2,7.7,2.6,6.9,2.3,Virginica
3,6.0,2.9,4.5,1.5,Versicolor
4,6.8,2.8,4.8,1.4,Versicolor


Why is it important?
- The dataset should not have an artificial ordering.

### Basic column/row operations

In [ ]:
#mean over a column
df['sepal_length'].mean()

np.float64(5.843333333333334)

In [ ]:
#mean over rows
df.iloc[:,:3].mean(axis=1)

0      3.333333
1      3.100000
2      3.066667
3      3.066667
4      3.333333
         ...   
145    4.966667
146    4.600000
147    4.900000
148    5.000000
149    4.666667
Length: 150, dtype: float64

Using the same method, we can also compute min, max, median, sum, etc.

### Exercises

Let’s load the TITANIC dataset


### The Titanic dataset

The Titanic dataset is one of the most well-known open datasets.
It is often used for practicing data processing, statistical analysis, and machine learning.
The data contain information about passengers of the RMS Titanic, which sank on April 15, 1912,
and the goal is to analyze who survived the disaster.

Dataset characteristics
- Source: OpenML (based originally on Kaggle and other statistical publications)
- Number of rows: 1309 passengers
- Number of columns: 14 features

Dataset columns
- pclass = passenger class (1 = first, 2 = second, 3 = third class)
- sex = passenger gender (male/female)
- age = passenger age (some values are missing)
- sibsp = number of siblings/spouses aboard
- parch = number of parents/children aboard
- fare = ticket price (in pounds)
- embarked = port of embarkation (C=Cherbourg, Q=Queenstown, S=Southampton)
- home.dest = passenger home or destination
- survived = target variable (0 = did not survive, 1 = survived)

Interesting facts about Titanic survival:
- First-class passengers had a much higher survival rate than third-class passengers.
- Women had significantly higher chances of survival than men.
- Due to the "women and children first" rule, younger passengers were more likely to survive.

In [ ]:
# Load Titanic data using scikit-learn

from sklearn.datasets import fetch_openml
import pandas as pd

# Load the dataset from OpenML
titanic = fetch_openml(name="titanic", version=1, as_frame=True)

# Convert data to a Pandas DataFrame
df = titanic.data
df["survived"] = titanic.target  # Add the target variable (y)

In [ ]:
df

,pclass,name,sex,age,sibsp,parch,ticket,fare,cabin,embarked,boat,body,home.dest,survived
0,1,"Allen, Miss. Elisabeth Walton",female,29.0000,0,0,24160,211.3375,B5,S,2,NaN,"St Louis, MO",1
1,1,"Allison, Master. Hudson Trevor",male,0.9167,1,2,113781,151.5500,C22 C26,S,11,NaN,"Montreal, PQ / Chesterville, ON",1
2,1,"Allison, Miss. Helen Loraine",female,2.0000,1,2,113781,151.5500,C22 C26,S,NaN,NaN,"Montreal, PQ / Chesterville, ON",0
3,1,"Allison, Mr. Hudson Joshua Creighton",male,30.0000,1,2,113781,151.5500,C22 C26,S,NaN,135.0,"Montreal, PQ / Chesterville, ON",0
4,1,"Allison, Mrs. Hudson J C (Bessie Waldo Daniels)",female,25.0000,1,2,113781,151.5500,C22 C26,S,NaN,NaN,"Montreal, PQ / Chesterville, ON",0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1304,3,"Zabour, Miss. Hileni",female,14.5000,1,0,2665,14.4542,NaN,C,NaN,328.0,NaN,0
1305,3,"Zabour, Miss. Thamine",female,NaN,1,0,2665,14.4542,NaN,C,NaN,NaN,NaN,0
1306,3,"Zakarian, Mr. Mapriededer",male,26.5000,0,0,2656,7.2250,NaN,C,NaN,304.0,NaN,0
1307,3,"Zakarian, Mr. Ortin",male,27.0000,0,0,2670,7.2250,NaN,C,NaN,NaN,NaN,0


### Exercise 1
- Show what travel classes existed on the ship
- Show what genders were present among the passengers

In [ ]:
import pandas as pd
from sklearn.datasets import fetch_openml

titanic = fetch_openml(name='titanic', version=1, as_frame=True)
df = titanic.frame

print(df['pclass'].unique())
print(df['sex'].unique())

[1 2 3]
['female', 'male']
Categories (2, object): ['female', 'male']


### Exercise 2
- Count how many women and men traveled on the ship
- Show (in %) how many women and men traveled on the ship
- Count, by travel class, how many women and men traveled on the ship

In [ ]:
import pandas as pd
from sklearn.datasets import fetch_openml

titanic = fetch_openml(name='titanic', version=1, as_frame=True)
df = titanic.frame

gender_counts = df['sex'].value_counts()
print(gender_counts)
print(gender_counts / len(df) * 100)
print(df.groupby(['pclass', 'sex']).size())

sex
male      843
female    466
Name: count, dtype: int64
sex
male      64.400306
female    35.599694
Name: count, dtype: float64
pclass  sex   
1       female    144
        male      179
2       female    106
        male      171
3       female    216
        male      493
dtype: int64


/tmp/ipykernel_10115/1396402409.py:10: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['pclass', 'sex']).size())


### Exercise 3
- Create a new column where you build categories from the AGE column:
  "child (<18), adult (18–65), senior (>65)"
- Count how many women and men traveled on the ship by age group
- Show (in %) how many women and men traveled on the ship by age group
- Count by travel class / age group how many women and men traveled on the ship

In [ ]:
import pandas as pd
from sklearn.datasets import fetch_openml

titanic = fetch_openml(name='titanic', version=1, as_frame=True)
df = titanic.frame

df['age_group'] = pd.cut(df['age'], bins=[0, 18, 65, 120], labels=['child', 'adult', 'senior'])
print(df.groupby(['age_group', 'sex']).size())
print(df.groupby(['age_group', 'sex']).size().groupby(level=0).apply(lambda x: x / x.sum() * 100))
print(df.groupby(['pclass', 'age_group', 'sex']).size())

age_group  sex   
child      female     92
           male      101
adult      female    295
           male      548
senior     female      1
           male        9
dtype: int64
age_group  age_group  sex   
child      child      female    47.668394
                      male      52.331606
adult      adult      female    34.994069
                      male      65.005931
senior     senior     female    10.000000
                      male      90.000000
dtype: float64
pclass  age_group  sex   
1       child      female     13
                   male        8
        adult      female    119
                   male      138
        senior     female      1
                   male        5
2       child      female     21
                   male       21
        adult      female     82
                   male      135
        senior     female      0
                   male        2
3       child      female     58
                   male       72
        adult      female     94
  

/tmp/ipykernel_10115/3050129027.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['age_group', 'sex']).size())
/tmp/ipykernel_10115/3050129027.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['age_group', 'sex']).size().groupby(level=0).apply(lambda x: x / x.sum() * 100))
/tmp/ipykernel_10115/3050129027.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['age_gro

### Exercise 4
The "survived" column shows whether someone survived the disaster.
- Show (in %) how many women and men survived
- Show by age group (in %) how many women and men survived
- Show by travel class (in %) how many passengers survived
- Show by travel class and gender (in %) how many passengers survived

In [ ]:
import pandas as pd
from sklearn.datasets import fetch_openml

titanic = fetch_openml(name='titanic', version=1, as_frame=True)
df = titanic.frame

df['survived'] = df['survived'].astype(int)
df['age'] = pd.to_numeric(df['age'], errors='coerce')

df['age_group'] = pd.cut(df['age'], bins=[0, 18, 65, 120], labels=['child', 'adult', 'senior'])

print(df.groupby('sex')['survived'].mean() * 100)
print(df.groupby(['age_group', 'sex'])['survived'].mean() * 100)
print(df.groupby('pclass')['survived'].mean() * 100)
print(df.groupby(['pclass', 'sex'])['survived'].mean() * 100)

sex
female    72.746781
male      19.098458
Name: survived, dtype: float64
age_group  sex   
child      female     68.478261
           male       31.683168
adult      female     77.288136
           male       18.613139
senior     female    100.000000
           male       11.111111
Name: survived, dtype: float64
pclass
1    61.919505
2    42.960289
3    25.528914
Name: survived, dtype: float64
pclass  sex   
1       female    96.527778
        male      34.078212
2       female    88.679245
        male      14.619883
3       female    49.074074
        male      15.212982
Name: survived, dtype: float64


/tmp/ipykernel_10115/2881678651.py:12: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby('sex')['survived'].mean() * 100)
/tmp/ipykernel_10115/2881678651.py:13: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['age_group', 'sex'])['survived'].mean() * 100)
/tmp/ipykernel_10115/2881678651.py:15: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['pclass', 'sex'])['survived'].mean()

### Exercise 5
The "survived" column shows whether someone survived the disaster.
- Show the average age of women and men
- Compare the age distribution of survivors vs non-survivors
- Compare the age distribution of survivors vs non-survivors by gender
- Compare the age distribution of survivors vs non-survivors by age group

In [ ]:
import pandas as pd
from sklearn.datasets import fetch_openml

titanic = fetch_openml(name='titanic', version=1, as_frame=True)
df = titanic.frame


df['survived'] = df['survived'].astype(int)
df['age'] = pd.to_numeric(df['age'], errors='coerce')

df['age_group'] = pd.cut(df['age'], bins=[0, 18, 65, 120], labels=['child', 'adult', 'senior'])

print(df.groupby('sex')['age'].mean())
print(df.groupby('survived')['age'].describe())
print(df.groupby(['survived', 'sex'])['age'].describe())
print(df.groupby(['survived', 'age_group'])['age'].describe())

sex
female    28.687071
male      30.585233
Name: age, dtype: float64
          count       mean        std     min   25%   50%   75%   max
survived                                                             
0         619.0  30.545369  13.922539  0.3333  21.0  28.0  39.0  74.0
1         427.0  28.918228  15.061481  0.1667  20.0  28.0  38.0  80.0
                 count       mean        std     min   25%   50%    75%   max
survived sex                                                                 
0        female   96.0  25.255208  13.476880  1.0000  18.0  24.5  31.25  63.0
         male    523.0  31.516412  13.796345  0.3333  22.0  29.0  40.00  74.0
1        female  292.0  29.815354  14.769278  0.1667  20.0  28.5  39.00  76.0
         male    135.0  26.977778  15.553884  0.4167  19.5  27.0  36.00  80.0
                    count       mean        std      min    25%    50%    75%  \
survived age_group                                                              
0        child      

/tmp/ipykernel_10115/3951724996.py:13: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby('sex')['age'].mean())
/tmp/ipykernel_10115/3951724996.py:15: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['survived', 'sex'])['age'].describe())
/tmp/ipykernel_10115/3951724996.py:16: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['survived', 'age_group'])['age'].describe())
